### PR TITLE
Multi-group Mode Implementation of Sparse Scattering Matrices and  Tallying Bug Fixes

### PR DESCRIPTION
**Pull request by nelsonag**
This PR Implements #602.  This was done, first by modifying scattdata to be a sparse format in which the full matrix input in the mgxs library is parses to find the minimum and maximum outgoing data, and then discards everything outside of these bounds.  After I did this, and put it on the nuclideMG and macroscopic data guys, I realized there is no reason we need to have different types for NuclideMG and Macroscopic; so, they are removed and replaced with type Mgxs.  This Mgxs type knows how to get its data from either an xml input file (like the nuclide did before) and how to create itself from a combination of other Mgxs objects (like the Macroscopic guy knew how to do with multiple nuclide data).

In the process of doing the above I noticed a few bugs which I fixed:
 1. A minor bug which probably affects no one, which can cause the bounding box for Legendre polynomial rejection sampling to not necessarily be the maximum (this would be the case for very forward peaked distributions).  In my tests this error didn't have a statistically significant effect, and this only shows up when using anisotropic scattering and choosing a non-default option.
 2. Tallies.  Basically, these were all hosed up.  The biggest reason why is I seem to have screwed these up is because I neglected to account for the effect on tallying of using material cross sections for reaction physics.  That is to say, multi-group particles don't interact with nuclides in the physics_mg module, they interact with the material as a whole.  So, what was affected by this bug?  Luckily, not the tallies one would mostly use from a multi-group solver: flux and fission rate tallies with one group.  That's why all our benchmark work above showed no issues.  I have now created a problem which tallies everything MG mode can tally (except for SCORE_TRANSPORT, I have spent no time on that, as you probably wont be surprised), and compared the resultant tallies to the theoretical results (i.e..... they better match the data inputted in to MGXS library).  I did this with and without survival biasing.  All match within statistics.

### CODE REVIEW

**Comment by nelsonag** [comment]
Sigh, tests are passing on my machine except for test_mgxs_library* (Pandas version issue? I was upgraded to 0.18 two days ago). No tests are passing on Travis.  Of course I find this out when nap time ends and so have to wait on fixing it...

**Comment by wbinventor** [comment]
@nelsonag what do the results true/error look like? I *doubt* this is a Pandas version issue...I think (hope) we are beyond all of that.

**Comment by wbinventor** [comment]
Also, if the issue is related to Pandas then that indicates that you are using a version *too new*! I think that Travis installs v0.17.X but I'm not sure. I didn't even know that v0.18 was out...now I have to go get it! 

I :blue_heart: Pandas

**Comment by nelsonag** [comment]
Ahh its the same as what @paulromano fixed in #610, results_true.dat has
"1" while results_error.dat has "1.0" (same for 0s).

**Comment by wbinventor** [comment]
Haha, looks like you are right - the same tests fail for me in #612 as well. I'll probably wait on @paulromano's PR to pull in the code needed to get around this since my PR isn't very time sensitive.

**Comment by nelsonag** [comment]
Ah yeah. I won't stress over the pandas related issues.  Now to find why all but one test is failing for me on Travis... How do you get Travis' openmc output again?

**Comment by wbinventor** [comment]
You need to add a `print(outstr)` to the `_get_results(...)` method of the `PyAPITestingHarness` class.

**Comment by nelsonag** [comment]
Well that was a mess.  After learning how to get the build output (which of course was really easy when you know what youre looking for), I learned that I am being burnt by a gfortran 4.6 issue.  Its especially odd because the offending line of code isnt anything new this time around.  Its moved to a different module, but all in all its the same thing.
```
[ 55%] Building Fortran object CMakeFiles/openmc.dir/src/mgxs_header.F90.o
/home/travis/build/mit-crpg/openmc/src/mgxs_header.F90:175.58:
    class(ScattDataContainer), allocatable :: scatter(:,:) ! scattering informa
.                                                          1
Fatal Error: Polymorphic array at (1) not yet supported
```

This is present in develop now (macroxs_header.F90), but gfortran has no issues with that there.  ScattDataContainer is your typical polymorphic container of class(ScattData).  The only important change to the ScattData extended types now is that I dont have real(8), allocatable members, but instead type(Jagged1D) members, which are just containers for real(8), allocatables to makefor jagged arrays in Fortran.

I have no fix at the moment, life hurts.

**Comment by wbinventor** [comment]
Pray for rain...I mean C++.

**Comment by nelsonag** [comment]
So Ubuntu 16.04 is out in a month.  Does anyone have any idea if Travis will make the switch as soon as it is out, or if there is a delay?

**Comment by paulromano** [comment]
Don't hold your breath @nelsonag

**Comment by nelsonag** [comment]
Too late. This PR has already taken my life.

**Comment by nelsonag** [comment]
This should work now, thankfully. The fix was not that bad and was kind of obvious, changed ``class(ScattDataContainer)`` to ``type(ScattDataContainer)``

**Comment by paulromano** [comment]
I'm seeing some weird behavior when tallying fission rate on a mesh in a [1D slab problem](https://www.dropbox.com/s/k9oj2n9gw0xxkze/toy.tar.xz?dl=0) with one-group cross sections. You would expect to see a cosine distribution, but instead I'm seeing the following:
flux

When I change the material to U-235 CE, I get:
flux_ce

Can you look into this @nelsonag?

I've also discovered bugs related to tallies on meshes/cells that line up with a reflective boundary, but that's another issue for another day.

**Comment by wbinventor** [comment]
Both @samuelshaner and myself have dealt with issues with mesh tallies which lead to very slightly errant results which may be consistent with your second point @paulromano (though I think the MG mesh tally issue is separate). In particular, we've seen cases where a very tiny tally will be made into a mesh cell when it in reality should be zero. An example would be "fission" or U-238 "absorption" in a water-filled guide tube in a PWR fuel assembly. In the 2-3 cases I've seen this come up, the tally is very small, on the order of machine precision, and it does not show up in the respective "distribcell" tally. I've been meaning to post an issue regarding this but have lost track of a good example to share to reproduce the error. Anyway, your post brought this to mind, so I'll try to dig something up to share.

**Comment by nelsonag** [comment]
Oooh, interesting.  Yes I will look in to it.

I'm seeing some weird behavior when tallying fission rate on a mesh in a 1D
slab problem <https://www.dropbox.com/s/k9oj2n9gw0xxkze/toy.tar.xz?dl=0>
with one-group cross sections. You would expect to see a cosine
distribution, but instead I'm seeing the following:
[image: flux]
<https://cloud.githubusercontent.com/assets/743095/13998970/e60dcd40-f107-11e5-8e34-f88c15776d98.png>

When I change the material to U-235 CE, I get:
[image: flux_ce]
<https://cloud.githubusercontent.com/assets/743095/13999045/610cbc2c-f108-11e5-81e6-9038fd9fb3b0.png>

Can you look into this @nelsonag <https://github.com/nelsonag>?

I've also discovered bugs related to tallies on meshes/cells that line up
with a reflective boundary, but that's another issue for another day.

—
You are receiving this because you were mentioned.
Reply to this email directly or view it on GitHub
<https://github.com/mit-crpg/openmc/pull/611#issuecomment-200521718>

**Comment by nelsonag** [comment]
Fixed!... Or at least understood!

Turns out the problem was far from converged.  I first noticed this was an issue when letting MPACT have a go at it and finding that even after 250 CMFD iterations, the fission source L2 norm was still 5 orders of magnitude away from the default convergence criteria.

So I went back to OpenMC, upped the inactive batches to 5,000 and added 5,000 active batches just to beat it to death.  I left the particles at 10,000/batch.  Lo and behold, I get the following figure (using the exact same inputs and script that @paulromano provided - save for the inactive/batches changes):

fission

Thank goodness, that was quite a scare.  I have been losing my mind in old age, but luckily not as much as I was fearing.

**Comment by paulromano** [comment]
Dope! Sorry for wasting your time. Now it's obvious -- total cross section of 0.3 cm^-1 and a region that is 400 cm wide... no wonder.

**Comment by nelsonag** [comment]
Yeah, I went to bed 2 nights ago thinking that was the issue, then I talked
myself out of it somehow :-(  Oh well.

But to be clear, the mesh tally issue is now considered separate, right?
@wbinventor does it seem that the mesh tallying issue you identified is
related to the one shown in the Feb 15 comment to #494 by @samuelshaner
showing errant fission mesh tally scores in the LRA problem?
@paulromano, your mesh tally issue seems separate from that, as that
location in the LRA is not close to a reflective boundary.

**Comment by paulromano** [comment]
I thought I had a separate issue, but now I'm having trouble reproducing it. If I do find something, I'll post a real issue.

**Comment by wbinventor** [comment]
@nelsonag - yes, the issue I observed seems to be related to the @samuelshaner's observation on Feb. 15. The only difference is that I've seen it in continuous energy mode and for various heterogeneous geometries while @samuelshaner saw it in MG mode for LRA. In each of our cases the error arose in the interior away from the boundaries. And it is not at all related to the changes In this PR, I simply brought it up based on @paulromano's comment.

**Comment by nelsonag** [comment]
@wbinventor got it, thanks!

File: src/mgxs_data.F90 [hunk 1]
```diff
@@ -118,17 +118,14 @@ subroutine read_mgxs()
...
           end select
 
           ! Now read in the data specific to the type we just declared
-          call nuclides_MG(i_nuclide) % obj % init(node_xsdata, energy_groups, &
-                                                   get_kfiss, get_fiss)
-
-          ! Keep track of what listing is associated with this nuclide
-          nuclides_MG(i_nuclide) % obj % listing = i_listing
+          call nuclides_MG(i_nuclide) % obj % init_file(node_xsdata, &
+               energy_groups,get_kfiss,get_fiss,max_order,i_listing)
```
Comment (paulromano) [inline src/mgxs_data.F90]: spaces after commas here

File: src/mgxs_data.F90 [hunk 2]
```diff
@@ -203,21 +177,15 @@ subroutine create_macro_xs()
...
       end select
-
-      call macro_xs(i_mat) % obj % init(mat, nuclides_MG, energy_groups, &
-                                        get_kfiss, get_fiss, max_order, &
-                                        scatt_type, legendre_mu_points, &
-                                        error_code, error_text)
-      ! Handle any errors
-      if (error_code /= 0) call fatal_error(trim(error_text))
+      call macro_xs(i_mat) % obj % combine(mat,nuclides_MG,energy_groups, &
+                                           max_order,scatt_type,i_mat)
```
Comment (paulromano) [inline src/mgxs_data.F90]: Please use a space after commas

File: src/particle_header.F90 [hunk 3]
```diff
@@ -232,7 +232,7 @@ subroutine create_secondary(this, uvw, type, run_CE)
     this % n_secondary = n
     this % secondary_bank(this % n_secondary) % E = this % E
     if (.not. run_CE) then
-      this % secondary_bank(this % n_secondary) % E = real(this % g, 8)
+      this % secondary_bank(this % n_secondary) % E = real(this % g,8)
```
Comment (paulromano) [inline src/particle_header.F90]: Comma should [remain](http://openmc.readthedocs.io/en/latest/devguide/styleguide.html#whitespace-in-expressions).

File: src/tally.F90 [hunk 4]
```diff
@@ -861,21 +901,24 @@ subroutine score_general_mg(p, t, start_index, filter_index, i_nuclide, &
           else
             score = p % last_wgt
           end if
+          if (i_nuclide > 0) then
+            score = score * atom_density * &
+                 nucxs % get_xs('total',p_g,UVW=p_uvw) / &
+                 matxs % get_xs('total',p_g,UVW=p_uvw)
```
Comment (paulromano) [inline src/tally.F90]: spaces after commas

File: src/tally.F90 [hunk 5]
```diff
@@ -861,21 +901,24 @@ subroutine score_general_mg(p, t, start_index, filter_index, i_nuclide, &
...
+          end if
 
         else
           if (i_nuclide > 0) then
-            associate (nuc => nuclides_MG(i_nuclide) % obj)
-              score = nuc % get_xs(p % g, 'total', UVW=p_uvw) * &
-                   atom_density * flux
-            end associate
+            score = nucxs % get_xs('total',p_g,UVW=p_uvw) * atom_density * flux
```
Comment (paulromano) [inline src/tally.F90]: ditto

File: src/tally.F90 [hunk 6]
```diff
@@ -886,175 +929,105 @@ subroutine score_general_mg(p, t, start_index, filter_index, i_nuclide, &
...
-            associate (nuc => nuclides_MG(i_nuclide) % obj)
-              score = nuc % get_xs(p % g, 'scatter', UVW=p_uvw) * &
-                   atom_density * flux
-            end associate
-          else
-            ! Get the scattering x/s (stored in % elastic)
-            score = material_xs % elastic * flux
+            score = score * atom_density * &
+                 nucxs % get_xs('scatter*f_mu/mult',p % last_g,p % g,UVW=p_uvw,MU=p % mu) / &
+                 matxs % get_xs('scatter*f_mu/mult',p % last_g,p % g,UVW=p_uvw,MU=p % mu)
```
Comment (paulromano) [inline src/tally.F90]: and again

File: src/tally.F90 [hunk 7]
```diff
@@ -886,175 +929,105 @@ subroutine score_general_mg(p, t, start_index, filter_index, i_nuclide, &
...
-                                                     p % last_uvw)
+          ! Note SCORE_SCATTER_*N not available for tracklength/collision.
+          if (i_nuclide > 0) then
+            score = atom_density * flux * &
+                 nucxs % get_xs('scatter/mult',p_g,UVW=p_uvw)
+          else
+            ! Get the scattering x/s and take away
+            ! the multiplication baked in to sigS
+            score = flux * &
+                 matxs % get_xs('scatter/mult',p_g,UVW=p_uvw)
```
Comment (paulromano) [inline src/tally.F90]: and again

File: src/tally.F90 [hunk 8]
```diff
@@ -886,175 +929,105 @@ subroutine score_general_mg(p, t, start_index, filter_index, i_nuclide, &
...
-          end associate
         else
-          score = score / &
-               macro_xs(p % material) % obj % get_xs(p % g, 'mult', &
-                                                     p % last_g, &
-                                                     p % last_uvw)
+          ! Note SCORE_SCATTER_*N not available for tracklength/collision.
+          if (i_nuclide > 0) then
+            score = atom_density * flux * &
+                 nucxs % get_xs('scatter/mult',p_g,UVW=p_uvw)
```
Comment (paulromano) [inline src/tally.F90]: and here

File: src/tally.F90 [hunk 9]
```diff
@@ -886,175 +929,105 @@ subroutine score_general_mg(p, t, start_index, filter_index, i_nuclide, &
...
-        ! the weight entering the collision as the estimator for the
-        ! reaction rate
-        score = p % last_wgt
+          ! Since we transport based on material data, the angle selected
+          ! was not selected from the f(mu) for the nuclide.  Therefore
+          ! adjust the score by the actual probability for that nuclide.
+          if (i_nuclide > 0) then
+            score = score * atom_density * &
+                 nucxs % get_xs('scatter*f_mu',p % last_g,p % g,UVW=p_uvw,MU=p % mu) / &
+                 matxs % get_xs('scatter*f_mu',p % last_g,p % g,UVW=p_uvw,MU=p % mu)
```
Comment (paulromano) [inline src/tally.F90]: some more here

**Comment by nelsonag** [comment]
Fixed the comma issues. Sorry about that @paulromano

File: src/scattdata_header.F90 [hunk 10]
```diff
@@ -8,36 +8,53 @@ module scattdata_header
...
-    ! p0 matrix on its own for sampling energy
-    real(8), allocatable :: energy(:,:) ! (Gout x Gin)
-    real(8), allocatable :: mult(:,:)   ! (Gout x Gin)
-    real(8), allocatable :: data(:,:,:) ! (Order/Nmu x Gout x Gin)
+    ! normalized p0 matrix on its own for sampling energy
+    type(Jagged1D), allocatable :: energy(:) ! (Gin % data(Gout))
+    ! nu-scatter multiplication (i.e. nu-scatt/scatt)
+    type(Jagged1D), allocatable :: mult(:)   ! (Gin % data(Gout))
+    ! Angular distribution
+    type(Jagged2D), allocatable :: dist(:)   ! (Gin % data(Order/Nmu x Gout)
```
Comment (paulromano) [inline src/scattdata_header.F90]: Might be worth putting a comment somewhere here noting that these arrays are not (necessarily) 1-indexed!

File: src/scattdata_header.F90 [hunk 11]
```diff
@@ -94,84 +116,169 @@ end subroutine scattdata_sample_
...
+      real(8), intent(in)                     :: mult(:,:)     ! Scatter Prod'n Matrix
+      real(8), intent(in)                     :: coeffs(:,:,:) ! Coefficients to use
+
+      real(8) :: dmu, mu, f, norm
+      integer :: imu, Nmu, gout, gin, groups, order
+      real(8), allocatable :: energy(:,:)
+      real(8), allocatable :: matrix(:,:,:)
+
+      groups = size(coeffs,dim=3)
+      order = size(coeffs,dim=1)
```
Comment (paulromano) [inline src/scattdata_header.F90]: spaces after commas

File: src/scattdata_header.F90 [hunk 12]
```diff
@@ -94,84 +116,169 @@ end subroutine scattdata_sample_
...
+
+      ! make a copy of coeffs that we can use to extract data and normalize
+      allocate(matrix(order,groups,groups))
+      matrix = coeffs
+
+      ! Get scattxs value
+      allocate(this % scattxs(groups))
+      ! Get this by summing the un-normalized P0 coefficient in matrix
+      ! over all outgoing groups
+      this % scattxs = sum(matrix(1,:,:),dim=1)
```
Comment (paulromano) [inline src/scattdata_header.F90]: here too

File: src/scattdata_header.F90 [hunk 13]
```diff
@@ -94,84 +116,169 @@ end subroutine scattdata_sample_
...
+      integer :: imu, Nmu, gout, gin, groups, order
+      real(8), allocatable :: energy(:,:)
+      real(8), allocatable :: matrix(:,:,:)
+
+      groups = size(coeffs,dim=3)
+      order = size(coeffs,dim=1)
+
+      ! make a copy of coeffs that we can use to extract data and normalize
+      allocate(matrix(order,groups,groups))
+      matrix = coeffs
```
Comment (paulromano) [inline src/scattdata_header.F90]: `matrix(:,:,:) = coeffs` is slightly safer. Technically F03+ is allowed to reallocate the LHS if you don't specify bounds like this.

File: src/scattdata_header.F90 (hunk 12 above)
Comment (paulromano) [inline src/scattdata_header.F90]: Also use `(:)` on LHS

File: src/scattdata_header.F90 [hunk 14]
```diff
@@ -94,84 +116,169 @@ end subroutine scattdata_sample_
...
-
-      groups = size(this % energy,dim=1)
+      allocate(this % max_val(groups))
+      ! Set dist values from matrix and initialize max_val
+      do gin = 1, groups
+        do gout = this % gmin(gin), this % gmax(gin)
+          this % dist(gin) % data(:,gout) = matrix(:,gout,gin)
+        end do
+        allocate(this % max_val(gin) % data(this % gmin(gin):this % gmax(gin)))
+        this % max_val(gin) % data = ZERO
```
Comment (paulromano) [inline src/scattdata_header.F90]: `data(:)`

File: src/scattdata_header.F90 [hunk 15]
```diff
@@ -94,84 +116,169 @@ end subroutine scattdata_sample_
...
-      integer :: imu, gin, gout, groups
+      integer :: imu, gin, gout, groups, order
       real(8) :: norm
-
-      groups = size(energy,dim=1)
+      real(8), allocatable :: energy(:,:)
+      real(8), allocatable :: matrix(:,:,:)
+
+      groups = size(coeffs,dim=3)
+      order = size(coeffs,dim=1)
```
Comment (paulromano) [inline src/scattdata_header.F90]: spaces after commas

File: src/scattdata_header.F90 [hunk 16]
```diff
@@ -94,84 +116,169 @@ end subroutine scattdata_sample_
...
+
+      ! make a copy of coeffs that we can use to extract data and normalize
+      allocate(matrix(order,groups,groups))
+      matrix = coeffs
```
Comment (paulromano) [inline src/scattdata_header.F90]: `matrix(:,:,:)`

File: src/scattdata_header.F90 [hunk 17]
```diff
@@ -94,84 +116,169 @@ end subroutine scattdata_sample_
...
+
+      ! Get scattxs value
+      allocate(this % scattxs(groups))
+      ! Get this by summing the un-normalized P0 coefficient in matrix
+      ! over all outgoing groups
+      this % scattxs = sum(sum(matrix(:,:,:),dim=1),dim=1)
```
Comment (paulromano) [inline src/scattdata_header.F90]: spaces after commas, use `this % scattxs(:)`

File: src/scattdata_header.F90 [hunk 18]
```diff
@@ -182,101 +289,138 @@ subroutine scattdatahistogram_init(this, order, energy, mult, coeffs)
...
 
-      if (order < 0) then
-        legendre_flag = .true.
-        this_order = -1 * order
-      else
-        legendre_flag = .false.
-        this_order = order
-      end if
+      groups = size(coeffs,dim=3)
+      order = size(coeffs,dim=1)
```
Comment (paulromano) [inline src/scattdata_header.F90]: spaces after commas

File: src/scattdata_header.F90 [hunk 19]
```diff
@@ -182,101 +289,138 @@ subroutine scattdatahistogram_init(this, order, energy, mult, coeffs)
...
 
-      groups = size(energy,dim=1)
+      ! make a copy of coeffs that we can use to extract data and normalize
+      allocate(matrix(order,groups,groups))
+      matrix = coeffs
```
Comment (paulromano) [inline src/scattdata_header.F90]: `matrix(:,:,:)`

File: src/scattdata_header.F90 [hunk 20]
```diff
@@ -285,58 +429,69 @@ end subroutine scattdatatabular_init
...
-      ! Adjust so interpolation works on the last bin if necessary
-      if (imu == size(this % data, dim=1)) then
-        imu = imu - 1
-      end if
+      if (gout < this % gmin(gin) .or. gout > this % gmax(gin)) then
+        f = ZERO
+      else
+        ! Find mu bin
+        if (mu == ONE) then
+          imu = size(this % fmu(gin) % data,dim=1)
```
Comment (paulromano) [inline src/scattdata_header.F90]: space before dim

File: src/scattdata_header.F90 [hunk 21]
```diff
@@ -285,58 +429,69 @@ end subroutine scattdatatabular_init
...
-      ! Adjust so interpolation works on the last bin if necessary
-      if (imu == size(this % data, dim=1)) then
-        imu = imu - 1
-      end if
+      if (gout < this % gmin(gin) .or. gout > this % gmax(gin)) then
+        f = ZERO
+      else
+        ! Find mu bin
+        if (mu == ONE) then
+          imu = size(this % fmu(gin) % data,dim=1) - 1
```
Comment (paulromano) [inline src/scattdata_header.F90]: space before dim

File: src/scattdata_header.F90 [hunk 22]
```diff
@@ -482,8 +637,85 @@ subroutine scattdatatabular_sample(this, gin, gout, mu, wgt)
...
+
+    pure function scattdata_get_matrix(this, req_order) result(matrix)
+      class(ScattData), intent(in) :: this          ! Scattering Object to work with
+      integer, intent(in)          :: req_order     ! Requested order of matrix
+      real(8), allocatable         :: matrix(:,:,:) ! Resultant matrix just built
+
+      integer :: order, groups, gin, gout
+
+      groups = size(this % energy)
+      order = min(req_order,size(this % dist(1) % data(:,1)))
```
Comment (paulromano) [inline src/scattdata_header.F90]: spaces

File: src/scattdata_header.F90 [hunk 23]
```diff
@@ -482,8 +637,85 @@ subroutine scattdatatabular_sample(this, gin, gout, mu, wgt)
...
+
+      allocate(matrix(order,groups,groups))
+      ! Initialize to 0; this way the zero entries in the dense matrix dont
+      ! need to be explicitly set, requiring a significant increase in the
+      ! lines of code.
+      matrix = ZERO
```
Comment (paulromano) [inline src/scattdata_header.F90]: `matrix(:,:,:)`

File: src/scattdata_header.F90 [hunk 24]
```diff
@@ -482,8 +637,85 @@ subroutine scattdatatabular_sample(this, gin, gout, mu, wgt)
...
+
+    pure function scattdatahistogram_get_matrix(this, req_order) result(matrix)
+      class(ScattDataHistogram), intent(in) :: this          ! Scattering Object to work with
+      integer, intent(in)                 :: req_order     ! Requested order of matrix
+      real(8), allocatable                :: matrix(:,:,:) ! Resultant matrix just built
+
+      integer :: order, groups, gin, gout
+
+      groups = size(this % energy)
+      order = min(req_order,size(this % dist(1) % data(:,1)))
```
Comment (paulromano) [inline src/scattdata_header.F90]: spaces between arguments

File: src/scattdata_header.F90 [hunk 25]
```diff
@@ -482,8 +637,85 @@ subroutine scattdatatabular_sample(this, gin, gout, mu, wgt)
...
+
+      allocate(matrix(order,groups,groups))
+      ! Initialize to 0; this way the zero entries in the dense matrix dont
+      ! need to be explicitly set, requiring a significant increase in the
+      ! lines of code.
+      matrix = ZERO
```
Comment (paulromano) [inline src/scattdata_header.F90]: `matrix(:,:,:)`

File: src/scattdata_header.F90 [hunk 26]
```diff
@@ -482,8 +637,85 @@ subroutine scattdatatabular_sample(this, gin, gout, mu, wgt)
...
+
+    pure function scattdatatabular_get_matrix(this, req_order) result(matrix)
+      class(ScattDataTabular), intent(in) :: this          ! Scattering Object to work with
+      integer, intent(in)                 :: req_order     ! Requested order of matrix
+      real(8), allocatable                :: matrix(:,:,:) ! Resultant matrix just built
+
+      integer :: order, groups, gin, gout
+
+      groups = size(this % energy)
+      order = min(req_order,size(this % dist(1) % data(:,1)))
```
Comment (paulromano) [inline src/scattdata_header.F90]: spaces between arguments. And do you really want to give a slice to `size(...)`? Couldn't you just specify which dimension?

File: src/scattdata_header.F90 [hunk 27]
```diff
@@ -482,8 +637,85 @@ subroutine scattdatatabular_sample(this, gin, gout, mu, wgt)
...
+
+      allocate(matrix(order,groups,groups))
+      ! Initialize to 0; this way the zero entries in the dense matrix dont
+      ! need to be explicitly set, requiring a significant increase in the
+      ! lines of code.
+      matrix = ZERO
```
Comment (paulromano) [inline src/scattdata_header.F90]: `matrix(:,:,:)`

File: src/tally.F90 [hunk 28]
```diff
@@ -886,175 +930,109 @@ subroutine score_general_mg(p, t, start_index, filter_index, i_nuclide, &
...
-                   atom_density * flux
-            end associate
-          else
-            ! Get the scattering x/s (stored in % elastic)
-            score = material_xs % elastic * flux
+            score = score * atom_density * &
+                 nucxs % get_xs('scatter*f_mu/mult', p % last_g, p % g, &
+                                UVW=p_uvw,MU=p % mu) / &
+                 matxs % get_xs('scatter*f_mu/mult', p % last_g, p % g, &
+                                UVW=p_uvw,MU=p % mu)
```
Comment (paulromano) [inline src/tally.F90]: Need a few spaces above

File: src/tally.F90 [hunk 29]
```diff
@@ -1089,38 +1069,30 @@ subroutine score_general_mg(p, t, start_index, filter_index, i_nuclide, &
...
+            score = p % last_wgt
           end if
-
+          if (i_nuclide > 0) then
+              score = score * atom_density * &
+                   nucxs % get_xs('fission',    p_g, UVW=p_uvw) / &
+                   matxs % get_xs('absorption', p_g, UVW=p_uvw)
+            else
+              score = score * &
+                   matxs % get_xs('fission',    p_g, UVW=p_uvw) / &
```
Comment (paulromano) [inline src/tally.F90]: Style guide also says avoid extraneous whitespace

File: src/tally.F90 [hunk 30]
```diff
@@ -1162,58 +1135,54 @@ subroutine score_general_mg(p, t, start_index, filter_index, i_nuclide, &
...
+            score = p % last_wgt
+          end if
+          if (i_nuclide > 0) then
+            score = score * atom_density * &
+                 nucxs % get_xs('kappa_fission', p_g, UVW=p_uvw) / &
+                 matxs % get_xs('absorption',    p_g, UVW=p_uvw)
+          else
+            score = score * &
+                 matxs % get_xs('kappa_fission', p_g, UVW=p_uvw) / &
+                 matxs % get_xs('absorption',    p_g, UVW=p_uvw)
```
Comment (paulromano) [inline src/tally.F90]: extraneous whitespace

**Comment by paulromano** [comment]
Hate it when github won't show me a diff. Here are comments on mgxs_header.F90:

- Spaces between procedure arguments in lines: 204, 262, 285, 553, 564, 587, 666, 1376, 1409, 1554, 1612, 1733
- line 295: Why `1 * groups`?
- line 332: Use `this % chi(:,:)`
- lines 378, 385: Use `temp_mult(:,:)`
- line 549: Use `this % total(:)`
- line 666: Use `this % nu_fission(:,:,:)`
- line 680: Use `temp_4d(:,:,:,:)` on LHS
- line 719: Use `this % fission(:,:,:)`
- line 732: Use `this % k_fission(:,:,:)`
- line 749: Use `this % absorption(:,:,:)`
- line 763: Use `this % mult(:,:,:,:)`
- line 839: Use `input_scatt(:,:,:,:,:)`
- line 954: Use `this % total(:,:,:)`
- line 1254: "youd" -> "you'd"
- lines 239, 429, 592, 598, 815, 1377, 1393, 1501, 1572, 1592, 1613, 1684: Not sure why the Error Message Has Each First Word Capitalized
- lines 1425-1444, 1455-1466, 1567, 1569, 1656-1674, 1687-1698: Use `(:)`, `(:,:)`, or `(:,:,:)` on LHS

**Comment by paulromano** [comment]
@nelsonag These changes all seem pretty reasonable. I like that the micros and macros now have a single type. Given how large the diff is, would you mind re-running some of the models that were used for "validation" when we reviewed the original MG pull request and share the results?

File: src/scattdata_header.F90 (hunk 26 above)
Comment (nelsonag) [inline src/scattdata_header.F90]: Yeah, and that would fix a bug I just realized in that the array may not be indexed by 1.

File: src/tally.F90 (hunk 29 above)
Comment (nelsonag) [inline src/tally.F90]: But.... but... alignment!

(fixed)

**Comment by nelsonag** [comment]
@paulromano: thanks for the thorough review! I have just resolved the above comments and will re-run some of those earlier models.

Thanks!

**Comment by nelsonag** [comment]
Ok, I re-ran some of those benchmark problems from before.  Here are the IHM, IHM-2group, 1d problem, C5G7 Pin, C5G7 2D, and C5G7 unrodded results compared with reference and previous results

Model         | Current OMC    |Previous OMC | Reference keff | Comments 
--------------|---------------------|--------------------|--------------------|-----------------
IHM 1grp    |  0.93750(NaN)  |   0.93750(NaN) |    0.9375 (analytical)   |  
IHM 2grp    |  1.72308(11)     |   1.72312(11)    |    1.72 (OpenMOC)     |
1d problem |  1.01091(13)     |   1.01067(13)    |    1.01062 (CE)           | known bug in legendre scattering fixed.  
C5G7 Pin    |  1.32566(5)      |   1.32560(5)      | N/A                              |
C5G7 2D    |   1.18655(5)     |    1.18646(5)     |    1.18655(80)              | I wish nailing it was anything besides just luck :-(
C5G7 3D    |   1.18373(9)     |1.18360(9)         |    1.18381(80)              |

**Comment by paulromano** [comment]
Thanks for the refactor/fixes!

**Pull request merged**